ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+
+
+/**
+ * Used to handle transactional writes to the command topic
+ */
+public class TransactionalProducerImpl implements TransactionalProducer {
+
+  private final TopicPartition commandTopicPartition;
+  private final String commandTopicName;
+  
+  private final Consumer<CommandId, Command> commandConsumer;
+  private final Producer<CommandId, Command> commandProducer;
+  private final Duration commandQueueCatchupTimeout;
+  private final CommandRunner commandRunner;
+
+  public TransactionalProducerImpl(
+      final String commandTopicName,
+      final CommandRunner commandRunner,
+      final Duration commandQueueCatchupTimeout,
+      final Map<String, Object> kafkaConsumerProperties,
+      final Map<String, Object> kafkaProducerProperties
+  ) {
+    this(
+        commandTopicName,
+        commandRunner,
+        commandQueueCatchupTimeout,
+        new KafkaConsumer<>(
+          Objects.requireNonNull(kafkaConsumerProperties, "kafkaConsumerProperties"),
+          InternalTopicJsonSerdeUtil.getJsonDeserializer(CommandId.class, true),
+          InternalTopicJsonSerdeUtil.getJsonDeserializer(Command.class, false)
+        ),
+        new KafkaProducer<>(
+          Objects.requireNonNull(kafkaProducerProperties, "kafkaProducerProperties"),
+          InternalTopicJsonSerdeUtil.getJsonSerializer(true),
+          InternalTopicJsonSerdeUtil.getJsonSerializer(false)
+        )
+    );
+  }
+
+  @VisibleForTesting
+  TransactionalProducerImpl(
+      final String commandTopicName,
+      final CommandRunner commandRunner,
+      final Duration commandQueueCatchupTimeout,
+      final Consumer<CommandId, Command> commandConsumer,
+      final Producer<CommandId, Command> commandProducer
+  ) {
+    this.commandTopicPartition = new TopicPartition(
+        Objects.requireNonNull(commandTopicName, "commandTopicName"),
+        0
+    );
+    this.commandQueueCatchupTimeout = 
+        Objects.requireNonNull(commandQueueCatchupTimeout, "commandQueueCatchupTimeout");
+    this.commandConsumer = Objects.requireNonNull(commandConsumer, "commandConsumer");
+    this.commandProducer = Objects.requireNonNull(commandProducer, "commandProducer");
+    this.commandTopicName = Objects.requireNonNull(commandTopicName, "commandTopicName");
+    this.commandRunner = Objects.requireNonNull(commandRunner, "commandRunner");
+    initialize();
+  }
+
+  private void initialize() {
+    commandConsumer.assign(Collections.singleton(commandTopicPartition));
+    commandProducer.initTransactions();
+  }
+
+  /** begins transaction */
+  public void begin() {
+    commandProducer.beginTransaction();
+    waitForConsumer();
+  }
+
+  public void waitForConsumer() {
```

Comment:
can this be made private? having it public makes it a little dangerous because if you call it outside of a transaction it doesn't really guarantee anything

Reply:
done